server/src/main/java/org/apache/iotdb/db/qp/Planner.java

```diff
@@ -139,6 +140,40 @@ public PhysicalPlan rawDataQueryReqToPhysicalPlan(
     return physicalGenerator.transformToPhysicalPlan(op, rawDataQueryReq.fetchSize);
   }
 
+  /** convert last data query to physical plan directly */
+  public PhysicalPlan lastDataQueryReqToPhysicalPlan(TSLastDataQueryReq req, ZoneId zoneId)
+      throws QueryProcessException, IllegalPathException {
+    List<String> paths = req.getPaths();
+    long time = req.getTime();
+
+    SelectOperator selectOp = new SelectOperator(SQLConstant.TOK_SELECT, zoneId);
+    FromOperator fromOp = new FromOperator(SQLConstant.TOK_FROM);
+    QueryOperator queryOp = new QueryOperator(SQLConstant.TOK_QUERY);
+
+    selectOp.setLastQuery();
+    selectOp.addSelectPath(new PartialPath(""));
+
+    for (String p : paths) {
+      PartialPath path = new PartialPath(p);
+      fromOp.addPrefixTablePath(path);
+    }
+
+    queryOp.setSelectOperator(selectOp);
+    queryOp.setFromOperator(fromOp);
+
+    PartialPath timePath = new PartialPath(TIME);
+
+    BasicFunctionOperator basicFunctionOperator =
+        new BasicFunctionOperator(SQLConstant.GREATERTHANOREQUALTO, timePath, Long.toString(time));
+    queryOp.setFilterOperator(basicFunctionOperator);
+
+    int maxDeduplicatedPathNum = Integer.MAX_VALUE - 1;
+    Operator op = logicalOptimize(queryOp, maxDeduplicatedPathNum);
+
+    PhysicalGenerator physicalGenerator = new PhysicalGenerator();
+    return physicalGenerator.transformToPhysicalPlan(op, req.fetchSize);
+  }
```

Comment:
A easy-fix problem is: you can see `rawDataQueryReqToPhysicalPlan`, the code logic is abstracted and very clear, we hope you can keep this code structure.

Reply:
Yes, I saw the master modify part of the code ,so i submit a new one